pandas/core/indexes/base.py

```diff
@@ -3696,6 +3696,13 @@ def get_indexer(
         tolerance=None,
     ) -> npt.NDArray[np.intp]:
         method = missing.clean_reindex_fill_method(method)
+        # After _maybe_cast_listlike_indexer, target elements which do not
+        # belong to some category are changed to NaNs
+        # Mask to track actual NaN values compared to inserted NaN values
+        try:
```

Comment:
this is likely not the right place for this (and would be a terrible performance hit), see @jbrockmendel suggestions

Reply:
instead of doing this, just define `orig_target = target` here, do the isna down near L3728 only if we actually need it.

Reply:
I have replied to those comments with what I have found. The reason for NaNs being assigned a non -1 index is because the position of NaNs is inferred from the Categorical Index. If the index does not have a NaN, this is -1. If it does, this value becomes >= 0. So at the moment, I refrained from trying to make changes in the internals, as I do not know the extent to which those changes might affect the codebase at large.
However, if there is a concrete direction to try, I could definitely work on it.

Reply:
Sure, that works